src/util/comparison.coffee

```diff
@@ -43,11 +43,47 @@ module.exports.greaterThanOrEquals = (a, b, precision = DateTime.Unit.MILLISECON
     else null
 
 module.exports.equivalent = equivalent = (a, b) ->
-  if a == null && b == null
-    return true
-  return a.hasMatch b if typeof a.hasMatch is 'function'
+  return true if not a? and not b?
+  return false unless a? and b?
+
+  return codesAreEquivalent(a, b) if isCode(a)
+
+  [aClass, bClass] = getClassOfObjects(a, b)
+  switch aClass
+    when '[object Array]'
+      return compareEveryItemInArrays(a, b, equivalent)
+    when '[object Object]'
+      return compareObjects(a, b, equivalent)
+
   return equals a, b
```

Comment:
I see that we are still falling back to `equals` as a last resort. I just want to make sure that those are **known** specialized cases. Since in the Spec it says `Note  that this operator will always return true or false, even if either or both of its arguments are null, or contain null components`. And the `equals` operator can return null in some cases. Specifically around date/time comparisons.

Reply:
`equals` can return null in two cases. The first case is
```coffeescript
return null unless a? and b?
```
This case will never happen because equivalent returns false before it falls through to `equals`:
```coffeescript
 return false unless a? and b?
```

The second case is 
```coffeescript
[aClass, bClass] = getClassOfObjects(a, b)
return false if aClass isnt bClass

switch aClass
...
when '[object Array]'
  return null if a.includes(null) or a.includes(undefined) or b.includes(null) or b.includes(undefined)
```
This case will also never fall through to `equals` because the same code appears in `equivalent` before the call to `equals`:

```coffeescript
[aClass, bClass] = getClassOfObjects(a, b)
switch aClass
  when '[object Array]'
    return compareEveryItemInArrays(a, b, equivalent)
```

Date/Time Comparisons, when they fall through to `equals`, can currently never return `null`.  If a future fix will end up with `equals` returning `null` in the case of `'[object Date]`, then another `when` statement in `equivalent` needs to be added to catch `[object Date]` before it falls through to equal.

Reply:
Writing that response made me realize that the line
```coffeescript
return false if aClass isnt bClass
```
isn't present in `equivalent`. 

I'll try to add a failing unit test then add that line.

Reply:
Added tests, which increased coverage, but wasn't able to get anything to break.